### ISSUE TITLE
Os or architecture for Arch is not properly detected

### ISSUE THREAD

**Issue by krafczyk** [LLNL/spack#2156]
I tried to install an older gcc on my arch machine for fun, and I ran into this nice problem:

```
spack install gcc@5.4.0
==> Installing gcc
==> Error: Specs gmp@6.1.1%gcc@6.2.1 arch=linux--x86_64^libsigsegv@2.10%gcc@6.2.1 arch=linux--x86_64^m4@1.4.17%gcc@6.2.1+sigsegv arch=linux--x86_64 and gmp@6.1.1%gcc@6.2.1 arch=linux-None-x86_64 have the same SHA-1 prefix!
```

You can clearly see that the os part of the triplet isn't detected properly. I have two triplets here, 'linux--x86_64' and 'linux-None-x86_64'.

**Comment by adamjstewart** [comment]
What is the output of:
```python
python -c 'import platform; print(platform.linux_distribution())'
```
I believe that's how we're currently determining the linux distribution.

**Comment by adamjstewart** [comment]
@tgamblin I added this to Spack v0.10 as I think it's a pretty critical bug. If Python's `platform` can't detect the Linux distribution, the least we should do is use "None" consistently.

**Comment by krafczyk** [comment]
Output follows:

```
python2 -c 'import platform; print(platform.linux_distribution())'
('', '', '')
```

interestingly, python3 gives a reasonable answer:

```
python3 -c 'import platform; print(platform.linux_distribution())'
('arch', '', '')
```

Actually I'm surprised by this.. I would've thought 2.7 would treat arch correctly. In any case, any reason why python3 isn't acceptable? if I run spack with python3 I get: 

`Spack requires Python 2.6 or 2.7. This is Python 3.5.2.`

**Comment by krafczyk** [comment]
Infact, there is a bug filed for this issue: https://bugs.python.org/issue20454

apparently they're thinking of deprecating `linux_distribution`.

**Comment by adamjstewart** [comment]
Spack currently doesn't support Python 3. There are plans to do this in the future, but first we would likely have to drop Python 2.6 support. We aren't really ready to do that as a lot of machines still don't come with Python 2.7 even. For example, I'm managing a RHEL 5 cluster that comes with Python 2.4.3 and a CentOS 6 cluster that comes with Python 2.6.6. The goal is that Spack should work out of the box with as few dependencies as possible.

I found a few interesting posts:
https://bugs.python.org/issue11678
https://github.com/ansible/ansible/issues/8653

It looks like `platform` was fixed in Python 3.3.0, but the fix hasn't been backported to Python 2.7 yet. Can you try running this command with Python 2?:
```python
python -c 'import platform; print(platform.linux_distribution(platform._supported_dists+("arch",)))'
```

**Comment by krafczyk** [comment]
That seems to work.

```
python2 -c 'import platform; print(platform.linux_distribution(supported_dists=platform._supported_dists+("arch",)))'
('arch', '', '')
```

But, does this mean that any unsupported dist will return as `('arch', '', '')`?

**Comment by adamjstewart** [comment]
I have no idea :smiley: 

But that's better than every unsupported dist crashing, in my mind. I'll put out a PR shortly.

**Comment by adamjstewart** [comment]
Can you try running the same command but replace "arch" with "unknown"? If it outputs `('unknown', '', '')` I think we'll know whether or not every unsupported arch will appear as "arch".

**Comment by krafczyk** [comment]
Yes, actually I just did that:

```
python2 -c 'import platform; print(platform.linux_distribution(supported_dists=platform._supported_dists+("other",)))'
('', '', '')
```

It seems that python is doing something more careful. In anycase, it would be good that the '' and 'None' issue doesn't happen for the next chap who has an unknown distro.

**Comment by adamjstewart** [comment]
I completely agree. My PR seems to work for Arch Linux, which is awesome. Ideally, if we are on an unsupported distro, it would use "unknown" or "none" consistently instead of None and '', but I'm not sure where that bug is getting introduced. I think #2157 is safe to merge, but someone else should take a look at why None and '' are being mixed up on unsupported distros.

**Comment by adamjstewart** [comment]
For the record, on Python 2.7.12, I'm seeing the following list for `platform._supported_dists`:
```python
('SuSE', 'debian', 'fedora', 'redhat', 'centos', 'mandrake', 'mandriva', 'rocks', 'slackware', 'yellowdog', 'gentoo', 'UnitedLinux', 'turbolinux')
```
On Python 3.5.1, I'm seeing:
```python
('SuSE', 'debian', 'fedora', 'redhat', 'centos', 'mandrake', 'mandriva', 'rocks', 'slackware', 'yellowdog', 'gentoo', 'UnitedLinux', 'turbolinux', 'arch', 'mageia')
```
The difference is that `arch` and `mageia` were added. Mageia seems pretty new, it's a 2010 fork of Mandriva, so we could probably add support for that if we wanted. I'm not sure how popular it is, but I don't want to add it to Spack unless we test it first.

**Comment by tgamblin** [comment]
Fixed by #1629.  See rationale there.
